File: steit-csharp/src/Collections/StateList.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;

using Steit.Codec;
using Steit.State;
using Steit.State.Event;

namespace Steit.Collections {
    public sealed class StateList<T> : ReadOnlyCollection<T>, IState {
        public Path Path { get; }

        // public StateList(Path? path = null, IList<T>? items = null) : base(items ?? new List<T>()) {
        public StateList(Path path = null, IList<T> items = null) : base(items ?? new List<T>()) {
            StateFactory.ValidateType(typeof(T));
            this.Path = path ?? Path.Root;
        }

        // public event EventHandler<FieldUpdateEventArgs<T, StateList<T>>>? OnUpdate;
        public event EventHandler<FieldUpdateEventArgs<T, StateList<T>>> OnUpdate;
        // public event EventHandler<ListPushEventArgs<T, StateList<T>>>? OnPush;
        public event EventHandler<ListPushEventArgs<T, StateList<T>>> OnPush;
        // public event EventHandler<ListPopEventArgs<T, StateList<T>>>? OnPop;
        public event EventHandler<ListPopEventArgs<T, StateList<T>>> OnPop;

        public void ClearUpdateHandlers() { this.OnUpdate = null; }
        public void ClearPushHandlers() { this.OnPush = null; }
        public void ClearPopHandlers() { this.OnPop = null; }

        // public static StateList<T> Deserialize(IReader reader, Path? path = null) {
        public static StateList<T> Deserialize(IReader reader, Path path = null) {
            // path ??= Path.Root;
            path = path ?? Path.Root;

            var items = new List<T>();
            var tag = 0U;

            while (!reader.EndOfStream()) {
                items.Add(reader.ReadValue<T>(path, tag++));
            }

            return new StateList<T>(path, items);
        }

        public WireType? GetWireType(UInt32 tag) {
            return StateFactory.IsStateType(typeof(T)) ? WireType.Sized : WireType.Varint;
        }

        // public IState? GetNested(UInt32 tag) {
        public IState GetNested(UInt32 tag) {
            return tag < this.Count ? this[(int) tag] as IState : null;
        }

        public void ReplaceAt(UInt32 tag, WireType wireType, IReader reader, bool shouldNotify) {
            if (tag >= this.Count) {
                throw new IndexOutOfRangeException();
            }

            var newItem = StateFactory.Deserialize<T>(reader, this.Path, tag);
            var oldItem = this[(int) tag];

            if (shouldNotify) {
                var args = new FieldUpdateEventArgs<T, StateList<T>>(tag, newItem, oldItem, this);
                this.OnUpdate?.Invoke(this, args);
            }

            this.Items[(int) tag] = newItem;
        }

        public void ReplayListPush(IReader reader) {
            var tag = (UInt32) this.Count;
            var item = StateFactory.Deserialize<T>(reader, this.Path, tag);

            var args = new ListPushEventArgs<T, StateList<T>>(tag, item, this);
            this.OnPush?.Invoke(this, args);

            this.Items.Add(item);
        }

        public void ReplayListPop() {
            if (this.Count <= 0) {
                throw new InvalidOperationException("Cannot pop from an empty `StateList`.");
            }

            var tag = (UInt32) this.Count - 1;

            var args = new ListPopEventArgs<T, StateList<T>>(tag, this[(int) tag], this);
            this.OnPop?.Invoke(this, args);

            this.Items.RemoveAt((int) tag);
        }

        public void ReplayMapRemove(UInt32 key) {
            throw new NotSupportedException();
        }
    }
}

File: steit-go/pkg/collections/statelist.go
package collections

import (
	"reflect"

	"github.com/axieinfinity/steit-go/pkg/codec"
	"github.com/axieinfinity/steit-go/pkg/eventhandler"
	"github.com/axieinfinity/steit-go/pkg/path"
	pathpkg "github.com/axieinfinity/steit-go/pkg/path"
	readerpkg "github.com/axieinfinity/steit-go/pkg/reader"
	statepkg "github.com/axieinfinity/steit-go/pkg/state"
)

var _ statepkg.IState = (*StateList)(nil)

type StateList struct {
	path     *pathpkg.Path
	items    []interface{}
	count    int
	OnUpdate eventhandler.EventHandler
	OnPush   eventhandler.EventHandler
	OnPop    eventhandler.EventHandler
}

func NewStateList(path *pathpkg.Path, items []interface{}) StateList {
	stateList := StateList{}

	if path != nil {
		stateList.path = path
	} else {
		stateList.path = pathpkg.Root
	}

	if len(items) > 0 {
		stateList.items = items
		stateList.count = len(items)
	}

	return stateList
}

func (s *StateList) GetItems() []interface{} {
	return s.items
}

func (s *StateList) GetCount() int {
	return s.count
}

func (s *StateList) ClearUpdateHandlers() {
	s.OnUpdate = nil
}

func (s *StateList) ClearPushHandlers() {
	s.OnPush = nil
}

func (s *StateList) ClearPopHandlers() {
	s.OnPop = nil
}

func Deserialize(reader readerpkg.IReader, path *pathpkg.Path) StateList {
	if path == nil {
		path = pathpkg.Root
	}

	var items []interface{}
	tag := uint32(0)

	for !reader.EndOfStream() {
		tag = tag + 1
		items = append(items, readerpkg.ReadValue(reader, path, tag))
	}

	return NewStateList(path, items)
}

func (s *StateList) GetWireType(tag uint32) *codec.WireType {
	if statepkg.IsStateType(reflect.TypeOf(s.items).Elem()) {
		c := codec.WireTypeSized
		return &c
	} else {
		c := codec.WireTypeVarint
		return &c
	}
}

func (s *StateList) GetNested(tag uint32) statepkg.IState {
	if int(tag) < s.count {
		if value, ok := s.items[tag].(statepkg.IState); !ok {
			panic("item not istate type")
		} else {
			return value
		}
		return s.items[tag].(statepkg.IState)
	} else {
		return nil
	}
}

func (s *StateList) ReplaceAt(tag uint32, wireType codec.WireType, reader readerpkg.IReader, shouldNotify bool) {
	if int(tag) >= s.count {
		panic("index out of range")
	}

	newItem := statepkg.Deserialize(reader, s.path, statepkg.DeserializeWithTag(tag))
	oldItem := s.items[tag]

	if shouldNotify {
		args := NewFieldUpdateEventArgs(tag, newItem, oldItem, s)
		if s.OnUpdate != nil {
			s.OnUpdate(s, args)
		}
	}

	s.items[tag] = newItem
}

func (s *StateList) ReplayListPush(reader readerpkg.IReader) {
	tag := uint32(s.count)
	item := statepkg.Deserialize(reader, s.path, statepkg.DeserializeWithTag(tag))

	args := NewListPushEventArgs(tag, item, s)
	if s.OnPush != nil {
		s.OnPush(s, args)
	}

	s.items = append(s.items, item)
}

func (s *StateList) ReplayListPop() {
	if s.count <= 0 {
		panic("Cannot pop from an empty `StateList`.")
	}

	tag := uint32(s.count - 1)

	args := NewListPopEventArgs(tag, s.items[tag], s)
	if s.OnPop != nil {
		s.OnPop(s, args)
	}

	s.items = append(s.items[:tag], s.items[tag+1:]...)
}

func (s *StateList) ReplayMapRemove(key uint32) {
	panic("not supported")
}

func (s *StateList) GetPath() *path.Path {
	return s.path
}
